### DISCUSSION TITLE
Get `*args` and `**kwargs` autocompletion but with modified types (wrapped in generic containers)

### DISCUSSION THREAD

**Discussion by Vince-LD** [python/typing#1646] [Q&A] [answered]
Hello,

I aplogize for the bad title formulation but I don't know how to phrase what I am looking for. I would be happy to change the title if you have suggestions.

I have a small value container class (`Field`) that allow me to get and set the value at multiple places/time the object as a reference to the value. These objects are passed to `Step` objects that have a run method.
 
The issue i am having is that I would like to create `FuncStep` class that wraps around a given function, exposes its `*args` and `**kwargs` but as a `TypeVarTuple[Field, ...]` instead of `TypeVarTuple`. I know this syntax is wrong, sorry for the abomination haha. The idea is to allow a user to have autocompletion for the arguments aswell as type checking. To make it clearer, I created a simple version of what It could look like. The code snippet is long, the intersting part is at the end.

```python
from typing import Callable, Generic, Optional, ParamSpec, TypeVar

T = TypeVar("T")
R = TypeVar("R")
P = ParamSpec("P")


class Field(Generic[T]):
    def __init__(self, value: T) -> None:
        self._value = value

    def get(self) -> T:
        return self._value

    def set(self, value: T):
        self._value = value


class Step:
    def run(self):
        ...

class FuncStep(Step, Generic[P]):
    def __init__(
        self,
        func: Callable[P, R],
        output: Field[Optional[R]],
        *args: P.args, # This should allow to use `func` arg names but wrapped in Field objects
        **kwargs: P.kwargs, # This should allow to use `func` arg names but wrapped in Field objects
    ) -> None:
        self.func = func
        self.output = output
        self.args = args 
        self.kwargs = kwargs
        super().__init__()

    def run(self):
        args = [arg.get() for arg in self.args]
        kwargs = {key: arg.get() for key, arg in self.kwargs.items()}
        self.output.set(self.func(*args, **kwargs))


def test_func(a: int, b: int) -> int:
    return a + b


if __name__ == "__main__":
    # These fields are Field[int]
    a_field, b_field, out_field = Field(1), Field(2), Field[Optional[int]](None)

    #I would like this FuncStep to await the parameters output: Field[int], a: Field[int] and b: Field[int]
    wrap_obj = FuncStep(test_func, output=out_field, a=a_field, b=b_field)
    wrap_obj.run() # This runs without error
    print(wrap_obj.output.get()) # We get the the right value

    # This is correct for the typechecker but doesn't work and doesn't allow me to 
    # modify the after the init for anywhere else in the program
    wrap_obj = FuncStep(test_func, output=out_field, a=8, b=7)
    wrap_obj.run()
    print(wrap_obj.output.get())
```

I guess a workaround solution would be to "cast" the `Field[T]` into the underlying `T` type so the type checker accepts `Field[int]` as input as it if was an `int` but I have no idea how to do that... As an example, I modified the Field class to have a cast property.

```python
class Field(Generic[T]):
    def __init__(self, value: T) -> None:
        self._value = value

    def get(self) -> T:
        return self._value

    def set(self, value: T):
        self._value = value

    @property
    def cast(self) -> T:
        return cast(T, self)

...

if __name__ == "__main__":
    # This works but is ugly, I would like not to have to use  the `.cast` property
    wrap_obj = FuncStep(test_func, output=out_field, a=a_field.cast, b=b_field.cast)
```

Thank you for your help!

**Comment by Daverball** [comment, marked as answer]
This sort of thing requires something like the previously proposed `Map` construct. Unfortunately no concrete proposal or PEP has come out of it yet, so the thing you're doing right now is pretty much the only thing you can do. SQLAlchemy used a very similar workaround before `TypeVarTuple` was available for their `Row` type, which had a `value` property that just returned `self` but cast it from `Row[T]` to `T`.

**Reply by Vince-LD** [reply]
Thank your for your answer. It's unfortunate there is no solution. However, I'm happy to find out my solution is somewhat standard. What I did is created a property called `T` that returns `self`.
I really hope something like `Map`  becomes available as it is a common pattern, especially in wrappers...
